RFCs/0009-Fn::ForEach.md

```diff
@@ -0,0 +1,1461 @@
+# Looping functionality for CFN Templates
+
+* **Original Author(s):**: @arthurboghossian
+* **Tracking Issue**: [Tracking Issue: Fn::ForEach](https://github.com/aws-cloudformation/cfn-language-discussion/issues/9)
+
+## Summary
+
+AWS CloudFormation is introducing a new `Fn::ForEach` intrinsic function. Customers can use this new feature to reduce verbosity of their CloudFormation templates, and improve their readability. This feature allows customers to declare multiple instances of similar resource type (ex. Subnets, VPCs) with a few lines of code. Customers can use this feature in Conditions, Resource, and Output sections of their template. Customers can declare any existing and future intrinsic functions such as `Fn::If`, `Fn::Join`, and more in `Fn::ForEach` intrinsic function. See [Examples](#examples) section for detailed use cases of `Fn::ForEach`, and sample templates with and without `Fn::ForEach` intrinsic function.
+
+## Motivation
+
+In a CloudFormation template, a single resource configures into one infrastructure object. A template, therefore, can become verbose when a customer manually declares similar resources. For example, customers want to declare a pool of `AWS::EC2::Instance` with the same configurations but different instance types, or `AWS::S3::Bucket NotificationConfiguration` with different `AWS::SNS::Topic`. Today, customers have to copy/paste the same lines of code with minor differences in resource properties.
+
+## Solution
+
+Customers can declare `Fn::ForEach` to iterate over a list of collections to generate desired fragments (ex. Resources).
+
+### JSON
+
+```yaml
+{
+    "Fn::ForEach::<UniqueLoopName>": [
+        "Identifier",
+        ["Value1", "Value2"], ## Collection
+        {
+            "OutputKey": "OutputValue" ## Ex: {"OutputKeyName${Identifier}": "OutputValue"}
```

Comment:
Using only a single item in the object could be read as only _permitting_ one item; it might be useful to include two items, or a comment indicating more items are possible